File: 第七章_二叉树/二叉树(2)/二叉树的序列化练习题/TreeToString.cpp
    /*
    struct TreeNode {
        int val;
        struct TreeNode *left;
        struct TreeNode *right;
        TreeNode(int x) :
                val(x), left(NULL), right(NULL) {
        }
    };*/
 
class TreeToString {
public:
    string toString(TreeNode* root) {
        string s;
        return PreOrderSerize(root,s);
    }
    //ÕûĞÍ×ª×Ö·û´®
    string int_tostring(int n)
    {
        string s;
        char a='0';
        while(n>0)
        {
            char ch=a+n%10;
            s=ch+s;
            n/=10;
        }
        return s;
    }
    //µİ¹éÏÈĞò±éÀú¶ş²æÊ÷
    string PreOrderSerize(TreeNode* root,string &s)
    {
        if(root==NULL)
        {
            s+="#!";
            return s;
        }
        string val=int_tostring(root->val);
        s+=val;
        s+="!";
        PreOrderSerize(root->left,s);
        PreOrderSerize(root->right,s);
 
        return s;
    }
};

File: 第七章_二叉树/二叉树(2)/二叉树的序列化练习题/TreeToString.java
import java.util.Stack;

public class TreeNode {
	int val = 0;
	TreeNode left = null;
	TreeNode right = null;

	public TreeNode(int val) {
		this.val = val;
	}
}

public class TreeToString {
	public String toString(TreeNode root) {

		if (root == null) {
			return "#!";
		}
		Stack<TreeNode> stack = new Stack<>();
		stack.push(root);
		StringBuilder builder = new StringBuilder();

		while (!stack.isEmpty()) {
			TreeNode cur = stack.peek();
			stack.pop();
			if (cur == null) {
				builder.append("#!");
				continue;
			} else {
				builder.append(cur.val);
				builder.append("!");
			}

			stack.push(cur.right);

			stack.push(cur.left);

		}

		return builder.toString();
	}
}